hazelcast/src/main/java/com/hazelcast/jet/impl/JobProxy.java

```diff
@@ -118,16 +119,24 @@ protected Address findLightJobCoordinator() {
 
     @Override
     protected CompletableFuture<Void> invokeSubmitJob(Object jobDefinition, JobConfig config) {
+        boolean serialize = true;
         if (isLightJob()) {
             if (jobDefinition instanceof DAG) {
-                ((DAG) jobDefinition).lock();
+                DAG dag = (DAG) jobDefinition;
+                dag.lock();
+                serialize = StreamSupport.stream(dag.spliterator(), false)
+                        .anyMatch(vertex -> vertex.getMetaSupplier().isStateful());
```

Comment:
`DAG.iterator()` is not a simple iteration. It uses `topologicalSort`. We should use something more lightweight here